src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+//
+// Parameters:
+// sizeKnown - if true, size is valid and needleLen invalid.
+// size - the size of the needle.  Pass 0 if unknown at compile time
+// noMatch - label bound outside to jump to if there is no match
+// loop_top - label bound inside this helper that should be branched to
+//            for additional comparisons.
+// eq_mask - The bit mask returned that holds the result of the comparison
+// hsPtrRet - This will hold the place within the needle where a match is found
+//            This is modified
+// needleLen - The length of the needle
+// ae - Argument encoding
+// _masm - Current MacroAssembler instance pointer
+//
+// On entry:
+//
+//  rbx: haystack
+//  rcx: k
+//  rdx: junk
+//  rsi: n
+//  rdi: haystack
+//  r10: n - k
+//  r12: k
+//  r13: junk
+//  r14: needle
+//  rbp: -1
+//  XMM_BYTE_0 - first element of needle broadcast
+//  XMM_BYTE_K - last element of needle broadcast
+
+static void big_case_loop_helper(bool sizeKnown, int size, Label &noMatch, Label &loop_top, Register eq_mask,
+                                 Register hsPtrRet, Register needleLen, StrIntrinsicNode::ArgEncoding ae,
+                                 MacroAssembler *_masm) {
+  Label temp;
+
+  assert_different_registers(eq_mask, hsPtrRet, needleLen, rdi, r15, rdx, rsi, rbx, r14, r10);
+
+  const Register needle = r14;
+  const Register haystack = rbx;
+  const Register hsLength = rsi;
+  const Register last = rdi;
+  const Register temp1 = r15;
+  const Register temp2 = rdx;
+
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+
+  assert_different_registers(eq_mask, hsPtrRet, needleLen, haystack, needle, hsLength, last);
+  //  hsLength, termAddr, last);
+
+  if (isU && (size & 1)) {
+    __ emit_int8(0xcc);
+    return;
+  }
+
+  broadcast_additional_needles(sizeKnown, size, NUMBER_OF_NEEDLE_BYTES_TO_COMPARE, needle, needleLen, temp1, isUU, isUL,
+                               _masm);
+
+  __ movq(r11, -1);
+
+  // Read 32-byte chunks at a time until the last 32-byte read would go
+  // past the end of the haystack.  Then, set the final read to read exactly
+  // the number of bytes in the haystack.
+  // For example, if haystack length is 45 and needle length is 13, the compares
+  // will read the following bytes:
+  //
+  //  First compare          Last compare
+  //   [  0 : 31]            [12 : 43]
+  // Next compare will go past end of haystack ([32:63])
+  // Adjust so final read is:
+  //   [  1 : 32]            [13 : 44]
+
+  __ movq(hsPtrRet, haystack);
+  // Assume r10 is n - k
+  __ leaq(last, Address(haystack, r10, Address::times_1, isU ? -30 : -31));
+  __ jmpb(temp);
+
+  __ align(16);
+  __ bind(loop_top);
+  __ cmpq(hsPtrRet, last);
+  __ je(noMatch);
+  __ addq(hsPtrRet, 32);
+
+  __ cmpq(hsPtrRet, last);
+  __ cmovq(Assembler::aboveEqual, hsPtrRet, last);
```

Comment:
cmovq is expensive, better sequence would be:

__ cmpq(hsPtrRet, last);
__ jb_b(temp);
__ movq(hsPtrRet, last);

Reply:
Done